### PR TITLE
Add geometry from elements and attribs to Event Analytics

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
- DHIS2-3351
- Modifed Event Analytics table generation process to extract the Geometry
column from Data Elements and Tracked Entity Attributes of type Org Unit
- Minor refactoring
- Added unit tests

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/JdbcEventAnalyticsTableManager.java [hunk 1]
```diff
@@ -231,119 +229,156 @@ protected void populateTable( AnalyticsTableUpdateParams params, AnalyticsTableP
             }
         }
 
-        List<OrganisationUnitLevel> levels =
-            organisationUnitService.getFilledOrganisationUnitLevels();
+        // Wrapper object for Program, required to avoid errors with Hibernate and non thread-safe Session
 
-        List<OrganisationUnitGroupSet> orgUnitGroupSets =
-            idObjectManager.getDataDimensionsNoAcl( OrganisationUnitGroupSet.class );
+        DetachedProgram detachedProgram = new DetachedProgram(program);
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/JdbcEventAnalyticsTableManager.java]: Ideally we would avoid having a `DetachedProgram` class. There are ways to attach background jobs to the Hibernate session.

Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/JdbcEventAnalyticsTableManager.java]: I have tried using `Hibernate.initialize( program )` but I still stumble in the same concurrent session access error. I see two alternatives: 

1. remove Futures and access the `Program` object from  a single thread
2. Use this pretty awful approach (which works, but..)
```
 private void init( Program program )
    {
        program.getDataElements().size();
        program.getDataElementsWithLegendSet().size();
        program.getNonConfidentialTrackedEntityAttributes().size();
        program.getNonConfidentialTrackedEntityAttributesWithLegendSet().size();
    }
```

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/JdbcEventAnalyticsTableManager.java [hunk 2]
```diff
@@ -227,91 +228,138 @@ protected void populateTable( AnalyticsTableUpdateParams params, AnalyticsTableP
...
+
+    private String selectForInsert(TrackedEntityAttribute attribute, String fromType, String dataClause) {
+
+        return String.format("(select %s"
+                + " from trackedentityattributevalue where trackedentityinstanceid=pi.trackedentityinstanceid "
+                + "and trackedentityattributeid=" + attribute.getId() + dataClause + ")" + getClosingParentheses( fromType )
+                + " as " + quote( attribute.getUid() ), fromType );
+    }
+
+    private List<AnalyticsTableColumn> getColumnFromDataElementWithLegendSet(DataElement dataElement, String select, String dataClause) {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/JdbcEventAnalyticsTableManager.java]: Codestyle, space after (, ), brackets on new lines.

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/util/AnalyticsTableAsserter.java [hunk 3]
```diff
@@ -103,6 +107,18 @@ public void verify()
...
+
+        for ( String name : matchers.keySet() )
+        {
+            if ( !tableColumnMap.containsKey( name ) )
+            {
+                fail( "Column [" + name + "] is missing" );
+            }
+            else
+            {
+                matchers.get(name).accept(tableColumnMap.get( name ));
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/util/AnalyticsTableAsserter.java]: Code style

**Pull request merged**